File: Python/987. Vertical Order Traversal of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

import queue

class Solution:
    def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        levels = queue.Queue()
        nodes = queue.Queue()
        nodes.put(root)
        levels.put((0,0))
        out = dict()
        while not nodes.empty():
            node = nodes.get()
            level = levels.get()
            if level[0] not in out:
                out[level[0]] = dict()
            if level[1] not in out[level[0]]:
                out[level[0]][level[1]] = list()            
            out[level[0]][level[1]].append(node.val)
            if node.left:
                nodes.put(node.left)
                levels.put((level[0] - 1, level[1] + 1))
            if node.right:
                nodes.put(node.right)
                levels.put((level[0] + 1, level[1] + 1))
        ans = list()        
        for x in sorted(out):
            line = list()
            for y in sorted(out[x]):
                line.extend(sorted(out[x][y]))
            ans.append(line)
        return ans

File: 2021/987. Vertical Order Traversal of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {        
        if (root == NULL) return {};             
        
        queue<TreeNode*> nodeQueue;
        queue<int> vQueue;
        nodeQueue.push(root);
        vQueue.push(0);        
        
        vector<vector<TreeNode*>> hMap;
        unordered_map<TreeNode*, int> vMap;
        
        int minV = 0, maxV = 0, currV = 0, currH = -1;
        while (!nodeQueue.empty()) {
            currH++;
            hMap.push_back({});
            int size = nodeQueue.size();
            for (int i = 0; i < size; i++) {
                root = nodeQueue.front();
                nodeQueue.pop();
                currV = vQueue.front();
                vQueue.pop();                
                vMap[root] = currV;
                hMap[currH].push_back(root);

                minV = min(minV, currV);
                maxV = max(maxV, currV);

                if (root->left) {
                    nodeQueue.push(root->left);
                    vQueue.push(currV - 1);
                }
                if (root->right) {
                    nodeQueue.push(root->right);
                    vQueue.push(currV + 1);
                }
            }                        
        }
        
        vector<vector<int>> ans(maxV - minV + 1, vector<int>{});
        for (int i = 0; i < hMap.size(); i++) {
            sort(hMap[i].begin(), hMap[i].end(), 
                 [](TreeNode* l, TreeNode* r) { return l->val < r->val; });
            for (TreeNode* node : hMap[i]) {
                ans[vMap[node]-minV].push_back(node->val);
            }            
        }
        return ans;        
    }
};